kusto/kcsb.go

```diff
@@ -0,0 +1,354 @@
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD application and key.
+func (kcsb connectionStringBuilder) WithAadAppKey(appId string, appKey string, authorityID string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationClientId, appId)
+	requireNonEmpty(applicationKey, appKey)
+	requireNonEmpty(authorityId, authorityID)
+	kcsb.applicationClientId = appId
+	kcsb.applicationKey = appKey
+	kcsb.authorityId = authorityID
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD application using a certificate.
+func (kcsb connectionStringBuilder) WithAppCertificate(appId string, certificate string, thumprint string, sendCertChain bool, authorityID string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationCertificate, certificate)
+	requireNonEmpty(authorityId, authorityID)
+	kcsb.applicationClientId = appId
+	kcsb.authorityId = authorityID
+
+	kcsb.applicationCertificate = certificate
+	kcsb.applicationCertificateThumbprint = thumprint
+	kcsb.sendCertificateChain = sendCertChain
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will authenticate with AAD application and an application token.
+func (kcsb connectionStringBuilder) WithApplicationToken(appId string, appToken string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationToken, appToken)
+	kcsb.applicationToken = appToken
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will use existing authenticated az cli profile password.
+func (kcsb connectionStringBuilder) WithAzCli() connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.azcli = true
+	return kcsb
+}
+
+/*
+Creates a KustoConnection string builder that will authenticate with AAD application, using
+  an application token obtained from a Microsoft Service Identity endpoint. An optional user
+  assigned application ID can be added to the token.
+*/
+func (kcsb connectionStringBuilder) WithManagedServiceID(clientID string, resId string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.msiAuthentication = true
+	if !isEmpty(clientID) {
+		kcsb.managedServiceIdentity = clientID
+	} else if !isEmpty(resId) {
+		kcsb.managedServiceIdentity = resId
+	}
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) WithInteractiveLogin(clientID string, authorityID string, redirectURL string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	if !isEmpty(clientID) {
+		kcsb.applicationClientId = clientID
+	}
+	if !isEmpty(authorityID) {
+		kcsb.authorityId = authorityID
+	}
+	if !isEmpty(redirectURL) {
+		kcsb.redirectURL = redirectURL
+	}
+	kcsb.interactiveLogin = true
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) AttachClientOptions(options *azcore.ClientOptions) connectionStringBuilder {
```

Comment:
What's the purpose of client options?

Reply:
Basically [`ClientOptions`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore#ClientOptions) contains optional settings for a client's pipeline. Like cloud.Configuration, logging policy, retry policy, telemetry policy. 
Have added this to provide flexibility/ability to so that `ClientOptions` can be attached.